lib/galaxy/model/migrate/triggers/update_audit_table.py

```diff
@@ -0,0 +1,140 @@
+from galaxy.model.migrate.versions.util import execute_statements
+
+
+# function name prefix
+fn_prefix = "fn_audit_history_by"
+
+# map between source table and associated incoming id field
+trigger_config = {
+    'history_dataset_association': "history_id",
+    'history_dataset_collection_association': "history_id",
+    'history': "id",
+}
+
+
+def install(engine):
+    """Install history audit table triggers"""
+    sql = _postgres_install() if 'postgres' in engine.name else _sqlite_install()
+    execute_statements(engine, sql)
+
+
+def remove(engine):
+    """Uninstall history audit table triggers"""
+    sql = _postgres_remove() if 'postgres' in engine.name else _sqlite_remove()
+    execute_statements(engine, sql)
+
+
+# Postgres trigger installation
+
+
+def _postgres_remove():
+    """postgres trigger removal sql"""
+
+    sql = []
+    sql.append(f"DROP FUNCTION IF EXISTS {fn_prefix}_history_id() CASCADE;")
+    sql.append(f"DROP FUNCTION IF EXISTS {fn_prefix}_id() CASCADE;")
+
+    return sql
+
+
+def _postgres_install():
+    """postgres trigger installation sql"""
+
+    sql = []
+
+    # postgres trigger function template
+    # need to make separate functions purely because the incoming history_id field name will be
+    # different for different source tables. There may be a fancier way to dynamically choose
+    # between incoming fields, but having 2 triggers fns seems straightforward
+
+    def trigger_fn(id_field):
+        fn = f"{fn_prefix}_{id_field}"
+
+        return f"""
+            CREATE OR REPLACE FUNCTION {fn}()
+                RETURNS TRIGGER
+                LANGUAGE 'plpgsql'
+            AS $BODY$
+                BEGIN
+                    INSERT INTO history_audit (history_id, update_time)
+                    SELECT DISTINCT {id_field}, CURRENT_TIMESTAMP AT TIME ZONE 'UTC'
+                    FROM new_table
+                    WHERE {id_field} IS NOT NULL;
```

Comment:
```suggestion
                    VALUES (NEW.{id_field}, CURRENT_TIMESTAMP AT TIME ZONE 'UTC')
```

Reply:
@nsoranzo That's a mistake I believe. With a statement level trigger we may be inserting many history_id, NOW combinations, not just one row.

Reply:
That's right, sorry!